syntax = "proto3";

package linuxcnc;

enum HalPinType {
    TYPE_UNSPECIFIED = 0;
    TYPE_UNINITIALIZED = 1;
    TYPE_BIT = 2;
    FLOAT = 3;
    S32 = 4;
    U32 = 5;
    PORT = 6;
}

enum HalPinDir {
    DIR_UNSPECIFIED = 0;
    IN = 1;
    OUT = 2;
    IO = 3;
}

message HalPin {
    string componentName = 1;
    string name = 2;
    HalPinType type = 3;
}

message CreateHalPinRequest {
    string name = 1;
    HalPinType type = 2;
    HalPinDir dir = 3;
}

message HalComponent {
    string name = 1;
}

message GetComponentsRequest {

}

message GetComponentsResponse {
    repeated HalComponent components = 1;
}

message GetPinsResponse {
    repeated HalComponent components = 1;
}

message CreateComponentRequest {
    string name = 1;
}

message CreateFloatPinRequest {
    string componentName = 1;
    float value = 2;
}

service LinuxCnc {
    rpc GetComponents(GetComponentsRequest) returns (stream GetComponentsResponse);

    rpc CreateComponent (CreateComponentRequest) returns (HalComponent);

    rpc GetPins(HalComponent) returns (stream GetPinsResponse);
    
    rpc CreateFloatPin (CreateFloatPinRequest) returns (HalPin);

}